pkg/csi/service/vanilla/controller.go

```diff
@@ -917,8 +919,22 @@ func (c *controller) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequ
 					"failed to get VolumeID from volumeMigrationService for volumePath: %q", volumePath)
 			}
 		}
-		// Check if the volume contains CNS snapshots.
-		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot) {
+		if cnsVolumeType == common.UnknownVolumeType {
+			cnsVolumeType, err = common.GetCnsVolumeType(ctx, c.manager, req.VolumeId)
+			if err != nil {
+				if err.Error() == common.ErrNotFound.Error() {
+					// The volume couldn't be found during query, assuming the delete operation as success
+					return &csi.DeleteVolumeResponse{}, "", nil
+				} else {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to determine CNS volume type for volume: %q. Error: %+v", req.VolumeId, err)
+				}
+			}
+			volumeType = convertCnsVolumeType(ctx, cnsVolumeType)
+		}
+		// Check if the volume contains CNS snapshots only for block volumes.
+		if cnsVolumeType == common.BlockVolumeType &&
+			commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot) {
```

Comment:
I think the GetCnsVolumeType should be made inside the check of BlockVolumeSnapshot FSS. 

Even though it won't make any difference as BlockVolumeSnapshot is going to be GA'ed this release. But, if some customers would like to explicitly opt out BlockVolumeSnapshot feature. This change would still introduce the performance impact. We should avoid it.

Reply:
The change also solves an existing "TODO" https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/c4b0961a4fde24226b91264d9ca9a52f935652fe/pkg/csi/service/vanilla/controller.go#L945

I'd leave it this way, from this change onwards metrics for file volume delete would be reported correctly, instead of incorrectly putting it under block volume metrics.

Having said that, I don't think we should allow users to turn off GA'ed feature, it may lead to unexpected behavior. The FSS is for our testing and alpha/beta features, it shouldn't be used as a kill switch for GAed feature.